modules: cmsis: move _glue_ code to modules/cmsis

gmarull opened the pull request:
Move the CMSIS _glue_ code to modules/cmsis. Refer to commits for more details.

gmarull committed:
modules: cmsis: move to a folder

Move CMSIS module Kconfig options to a folder, in preparation for CMSIS
core header creation.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
modules: cmsis: move glue code to modules/cmsis

The CMSIS module glue code was part of arch/ directory. Move it to
modules/cmsis, and provide a single entry point for it: cmsis_core.h.
This entry header will include the right CMSIS header (M or A/R).

To make this change possible, CMSIS module Kconfig/CMake are declared as
external, allowing us to add a new Zephyr include directory.

All files including CMSIS have been updated.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
arch: arm: aarch32: barrier: remove redundant CMSIS ifdeffery

CMSIS header is now generic.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
cmsis: remove unnecessary includes

Some files included <cmsis_core.h> for nothing, delete it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
arch: arm: aarch32: mpu: remove redundant guard

CMSIS header is now generic.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

aescolar commented on modules/cmsis/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+add_subdirectory(${ZEPHYR_CURRENT_MODULE_DIR} cmsis)
+zephyr_include_directories(.)
```
Adding it like this unconditionally, it gets added to all builds.
So, for simulation, now we depend on the include path order. How things stand right now, this path is after the nrf52_bsim cmsis/ include, which means things did not blow up.
I'd add a ifdef ARM archs here or so, as this headers are the CMSIS Core glue.

gmarull replied:
the include path is unconditional, but same was before with the include found in `zephyr/arch/...`, right? I'm not sure I understand the problem here.

aescolar replied:
That before it was added only if you were building with the ARM archs. Now it is always.

gmarull replied:
As is today, cmsis _glue_ code resides in:

```
include/zephyr/arch/arm/aarch32/cortex_a_r/cmsis.h
include/zephyr/arch/arm/aarch32/cortex_m/cmsis.h
```

Which is always in the include path, no?

aescolar replied:
Only for ARM targets

gmarull replied:
ok, got the problem, bsim has its own `cmsis.h` in the path as well. should now be fixed with `cmsis_core.h` (also, include path is only added if actually makes sense, ie for M/A/R)

gmarull commented:
@SgrrZhf rebased

MaureenHelm commented:
Does this break out of tree SoCs?

gmarull commented:
> Does this break out of tree SoCs?

Yes, header can be kept with a warning though if preferred.

gmarull committed:
cmsis: add deprecated headers

To ease migration of out-of-tree code, re-introduce previous CMSIS
headers that when include generate a warning informing about the
replacement header.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
@MaureenHelm added https://github.com/zephyrproject-rtos/zephyr/pull/59855/commits/eab032750b2cc6c59d7db02d4f1516722c3f2ca5, it's a trivial case that will help.

gmarull commented:
last push: updated module SHAs
